clouddriver-ecs/src/test/java/com/netflix/spinnaker/clouddriver/ecs/cache/ApplicationCacheClientTest.java

```diff
@@ -0,0 +1,163 @@
+    clusterNames.put(ACCOUNT, Sets.newHashSet(SERVICE_NAME));
+
+    Application application = new EcsApplication(APP_NAME, appAttributes, clusterNames);
+
+    Map<String, Object> cacheAttributes = new HashMap<>();
+    cacheAttributes.put("name", application.getName());
+
+    when(cacheView.get(eq(ECS_APPLICATIONS.ns), eq(key), any()))
+        .thenReturn(new DefaultCacheData(key, cacheAttributes, relationships));
+
+    // When
+    Application retrievedApplication = client.getApplication(APP_NAME);
+
+    // Then
+    assertTrue(
+        "Expected the application to be " + application + " but got " + retrievedApplication,
+        application.equals(retrievedApplication));
+  }
+
+  @Test
+  public void shouldGetApplications() {
+    // Given
+    String key = Keys.getApplicationKey(APP_NAME);
+    String serviceKey = Keys.getServiceKey(ACCOUNT, REGION, SERVICE_NAME);
+
+    Map<String, String> appAttributes = new HashMap<>();
+    appAttributes.put("name", APP_NAME);
+
+    Map<String, Collection<String>> relationships = new HashMap<>();
+    relationships.put(SERVICES.ns, Sets.newHashSet(serviceKey));
+
+    Map<String, Set<String>> clusterNames = new HashMap<>();
+    clusterNames.put(ACCOUNT, Sets.newHashSet(SERVICE_NAME));
+
+    Application application = new EcsApplication(APP_NAME, appAttributes, clusterNames);
+
+    Map<String, Object> cacheAttributes = new HashMap<>();
+    cacheAttributes.put("name", application.getName());
+
+    when(cacheView.getAll(eq(ECS_APPLICATIONS.ns), any(), any()))
+        .thenReturn(Sets.newHashSet(new DefaultCacheData(key, cacheAttributes, relationships)));
+
+    // When
+    Set<Application> retrievedApplication = client.getApplications(false);
+
+    // Then
+    assertTrue(
+        "Expected the application to be " + application + " but got " + retrievedApplication,
+        Sets.newHashSet(application).equals(retrievedApplication));
+  }
+
+  @Test
+  public void shouldGetApplicationsExpanded() {
+    // Given
+    String key = Keys.getApplicationKey(APP_NAME);
+    String serviceKey = Keys.getServiceKey(ACCOUNT, REGION, SERVICE_NAME);
+
+    Map<String, String> appAttributes = new HashMap<>();
+    appAttributes.put("name", APP_NAME);
+
+    Map<String, Collection<String>> relationships = new HashMap<>();
+    relationships.put(SERVICES.ns, Sets.newHashSet(serviceKey));
+
+    Map<String, Set<String>> clusterNames = new HashMap<>();
+    clusterNames.put(ACCOUNT, Sets.newHashSet(SERVICE_NAME));
+
+    Application application = new EcsApplication(APP_NAME, appAttributes, clusterNames);
+
+    Map<String, Object> cacheAttributes = new HashMap<>();
+    cacheAttributes.put("name", application.getName());
+
+    when(cacheView.getAll(eq(ECS_APPLICATIONS.ns), any(), any()))
+        .thenReturn(Sets.newHashSet(new DefaultCacheData(key, cacheAttributes, relationships)));
+
+    // When
+    Set<Application> retrievedApplication = client.getApplications(true);
+
+    // Then
+    assertTrue(
```

Comment:
we're asserting it's the right application, but shouldn't we also assert on the expected service relationship(s) ?

Reply:
The application object contains the services as the cluster names, so it's actually comparing the entire application object not just the app name itself.

Reply:
oh I see we're comparing the HashSets - nvm!